### PR TITLE
[FIX]: Handle Error response from AI API

### PR DESCRIPTION
**Pull request by VipinDevelops**
## Linked Issue(s)

#546 

## Acceptance Criteria fulfillment

- [x]  Handle the AI API response in the backend, ensuring proper error handling.
- [x]  Handle the BFF response for DORA metrics ensuring proper error feedback to the frontend.

## Proposed changes (including videos or screenshots)
[Screencast from 2024-09-15 11-54-08.webm](https://github.com/user-attachments/assets/67f10eb8-2f2b-4c5e-a5d1-f155a177c1ed)

### CODE REVIEW

File: web-server/pages/api/internal/ai/dora_metrics.ts [hunk 1]
```diff
@@ -64,51 +64,83 @@ const postSchema = yup.object().shape({
...
+      error: error.message
+    });
+  }
 });
 
+function checkForErrors(responses: any): { status: string; message: string } {
+  let status = 'success';
+  let message = '';
+
+  for (const value of Object.values(responses)) {
```
Comment (e-for-eshaan) [inline web-server/pages/api/internal/ai/dora_metrics.ts:133]:
```ts
const errorResponse = Object.values(responses))
                      .find(res => res.status==='error')

const {status = 'success', message = ''} = errorResponse ?? {} 
// if `errorResponse` is not found, default it to an empty object
```

Comment (e-for-eshaan) [reply web-server/pages/api/internal/ai/dora_metrics.ts:133]: we rarely use `for` and `whiles`

Comment (e-for-eshaan) [reply web-server/pages/api/internal/ai/dora_metrics.ts:133]: in fact use an if statement
```ts
if(error message)
{
  return ...
}

return ...
```

File: web-server/pages/api/internal/ai/dora_metrics.ts [hunk 2]
```diff
@@ -64,51 +64,83 @@ const postSchema = yup.object().shape({
...
+    return res.status(200).send(simplifiedData);
+  } catch (error) {
+    return res.status(500).send({
+      message: 'Internal Server Error',
+      error: error.message
+    });
+  }
 });
 
+function checkForErrors(responses: any): { status: string; message: string } {
```
Comment (e-for-eshaan) [inline web-server/pages/api/internal/ai/dora_metrics.ts:129]: usually we use const func = () => {}

**Comment by e-for-eshaan** [comment]
@samad-yar-khan will submit the final review

File: web-server/pages/api/internal/ai/dora_metrics.ts [hunk 3]
```diff
@@ -125,14 +125,12 @@ endpoint.handle.POST(postSchema, async (req, res) => {
     });
   }
 });
-
-const checkForErrors = (responses: any): { status: string; message: string } => {
+const checkForErrors = (responses: { [key: string]: { status: string; message: string } }): { status: string; message: string } => {
```
Comment (e-for-eshaan) [inline web-server/pages/api/internal/ai/dora_metrics.ts:128]: use `Record`, not that what you have written is wrong, but consistent throughout the codebase

Comment (VipinDevelops) [reply web-server/pages/api/internal/ai/dora_metrics.ts:128]: Sure

File: backend/analytics_server/mhq/service/ai/ai_analytics_service.py [hunk 4]
```diff
@@ -73,28 +97,28 @@ def _fireworks_ai_fetch_completions(self, messages: List[Dict[str, str]]):
...
-            raise Exception(response.json())
-
-        return response.json()
-
-    def _fetch_completion(self, messages: List[Dict[str, str]]):
+        return self._handle_api_response(response)
 
+    def _fetch_completion(
+        self, messages: List[Dict[str, str]]
+    ) -> Union[Dict[str, str], Dict[str, Union[str, int]]]:
```
Comment (samad-yar-khan) [inline backend/analytics_server/mhq/service/ai/ai_analytics_service.py:104]: this seems over complicated

File: backend/analytics_server/mhq/service/ai/ai_analytics_service.py [hunk 5]
```diff
@@ -73,28 +97,28 @@ def _fireworks_ai_fetch_completions(self, messages: List[Dict[str, str]]):
...
-        raise Exception(f"Invalid AI provider {self._ai_provider}")
+        return {
+            "status": "error",
+            "message": f"Invalid AI provider {self._ai_provider}",
+        }
 
     def get_dora_metrics_score(
         self, four_keys_data: Dict[str, float]
-    ) -> Dict[str, str]:
+    ) -> Union[Dict[str, str], Dict[str, Union[str, int]]]:
```
Comment (samad-yar-khan) [inline backend/analytics_server/mhq/service/ai/ai_analytics_service.py:121]: although this serves our usecase, this return type feels like an overkill

Comment (samad-yar-khan) [reply backend/analytics_server/mhq/service/ai/ai_analytics_service.py:121]: and super hard to decipher

**Review by samad-yar-khan — CHANGES_REQUESTED** [review]

File: backend/analytics_server/mhq/service/ai/ai_analytics_service.py (hunk 4 above)
Comment (VipinDevelops) [reply backend/analytics_server/mhq/service/ai/ai_analytics_service.py:104]: i'll try to simplify it

File: backend/analytics_server/mhq/service/ai/ai_analytics_service.py (hunk 5 above)
Comment (VipinDevelops) [reply backend/analytics_server/mhq/service/ai/ai_analytics_service.py:121]: I will make it simpler TBH I don't like this either

**Review by samad-yar-khan — APPROVED** [review]
LGTM, but I think a lot of this can be made better by better exception handling, lets take those up in future PRs: https://flask.palletsprojects.com/en/2.3.x/errorhandling/

**Comment by e-for-eshaan** [comment]
@VipinDevelops merge this then!

**Pull request merged**